File: cses/08_string_algorithms/3138_all-palindromes/solution.py
import sys


def main():
    data = sys.stdin.buffer.read().split()
    if not data:
        return
    s = data[0]
    n = len(s)

    # Xâu 1-indexed với sentinel a[0] = -1 (khác mọi mã kí tự 0..25)
    # để bảo đảm chỉ số i - length[cur] - 1 luôn hợp lệ (>= 0).
    base = 97  # ord('a')
    a = [-1] * (n + 1)
    for i in range(1, n + 1):
        a[i] = s[i - 1] - base

    # Eertree: nút 0 = imaginary root (len = -1), nút 1 = empty root (len = 0).
    length = [-1, 0]
    link = [0, 0]
    trans = {}            # key = node * 26 + c  ->  nút con (palindrome c + P + c)
    last = 1              # palindrome-suffix dài nhất của tiền tố hiện tại
    ans = [0] * n

    # Đưa về biến cục bộ cho nhanh.
    _length = length
    _link = link
    _trans = trans

    # Thêm từng kí tự từ trái sang phải.
    for i in range(1, n + 1):
        c = a[i]
        # Tìm cur: suffix-palindrome dài nhất có thể mở rộng bằng kí tự c.
        cur = last
        while a[i - _length[cur] - 1] != c:
            cur = _link[cur]

        key = cur * 26 + c
        child = _trans.get(key)
        if child is not None:
            # Palindrome c + P(cur) + c đã tồn tại, chỉ cập nhật last.
            last = child
            ans[i - 1] = _length[child]
            continue

        # Tạo nút palindrome mới với độ dài length[cur] + 2.
        newlen = _length[cur] + 2
        if newlen == 1:
            sl = 1  # link tới empty root
        else:
            # Tìm suffix link cho nút mới bằng cách đi tiếp theo suffix link từ link[cur].
            tmp = _link[cur]
            while a[i - _length[tmp] - 1] != c:
                tmp = _link[tmp]
            sl = _trans[tmp * 26 + c]

        node = len(_length)
        _length.append(newlen)
        _link.append(sl)
        _trans[key] = node
        last = node
        # length[last] = độ dài palindrome-suffix dài nhất kết thúc tại i.
        ans[i - 1] = newlen

    sys.stdout.write(' '.join(map(str, ans)))
    sys.stdout.write('\n')


main()

File: cses/08_string_algorithms/3138_all-palindromes/solution.cpp
#include <array>
#include <iostream>
#include <string>
#include <vector>

// Mỗi nút của Eertree ứng với một palindrome phân biệt.
struct Node {
    int length;                  // độ dài palindrome
    int link;                    // suffix link: palindrome-suffix thật sự dài nhất
    std::array<int, 26> next{};  // transition: nút ứng với palindrome c + P + c

    Node(const int node_length, const int suffix_link)
        : length(node_length), link(suffix_link) {
        next.fill(-1);
    }
};

int main() {
    std::ios::sync_with_stdio(false);
    std::cin.tie(nullptr);

    std::string text;
    std::cin >> text;
    const int size = static_cast<int>(text.size());

    // Xâu 1-indexed với sentinel characters[0] = -1 (khác mọi mã kí tự 0..25)
    // để bảo đảm chỉ số index - length - 1 luôn hợp lệ, imaginary root luôn khớp.
    std::vector<int> characters(static_cast<std::size_t>(size + 1), -1);
    for (int index = 1; index <= size; ++index) {
        characters[static_cast<std::size_t>(index)] =
            text[static_cast<std::size_t>(index - 1)] - 'a';
    }

    // Hai gốc: nút 0 = imaginary root (len = -1), nút 1 = empty root (len = 0).
    std::vector<Node> tree;
    tree.reserve(text.size() + 2);
    tree.emplace_back(-1, 0);
    tree.emplace_back(0, 0);
    int last = 1;  // palindrome-suffix dài nhất của tiền tố hiện tại

    // Thêm từng kí tự từ trái sang phải.
    for (int index = 1; index <= size; ++index) {
        const int character = characters[static_cast<std::size_t>(index)];

        // Đi theo suffix link tìm current: suffix-palindrome dài nhất bọc thêm được kí tự.
        int current = last;
        while (characters[static_cast<std::size_t>(
                   index - tree[current].length - 1
               )] != character) {
            current = tree[current].link;
        }

        int child = tree[current].next[character];
        if (child == -1) {
            // Palindrome c + P(current) + c chưa tồn tại: tạo nút mới.
            const int new_length = tree[current].length + 2;
            int suffix_link = 1;
            if (new_length > 1) {
                // Tìm suffix link cho nút mới bằng cách tiếp tục đi từ link[current].
                int candidate = tree[current].link;
                while (characters[static_cast<std::size_t>(
                           index - tree[candidate].length - 1
                       )] != character) {
                    candidate = tree[candidate].link;
                }
                suffix_link = tree[candidate].next[character];
            }

            child = static_cast<int>(tree.size());
            tree.emplace_back(new_length, suffix_link);
            tree[current].next[character] = child;
        }
        last = child;

        // length[last] = độ dài palindrome-suffix dài nhất kết thúc tại index.
        if (index > 1) {
            std::cout << ' ';
        }
        std::cout << tree[last].length;
    }
    std::cout << '\n';
    return 0;
}
